OMPD 5 and limited OMPD support for cuda devices

manorom opened the pull request:
Aligns the OMPD interface with the draft specification of OpenMP 5 and adds limited OMPD support for cuda devices (thread handles, some thread states, parallel handels).

Commit:
Merge branch 'master' into ompd

manorom committed:
[OMPD] Fixing runtime ompd type references

manorom committed:
[OMPD] add cmake modules necessary to build gdb-wrapper

manorom committed:
[OMPD] align libompd and odb w/ new OpenMP spec

This commit aligns libompd and gdb-wrapper with a newer OpenMP spec.
* Some functionality was desiabled as it relies on information no longer
exposed via OMPD.
* The OMPD callback functions still have non-standard names (but
  signatures should be correct now)
* Cuda device initialization should work now, but not much other OMPD
  device funtionality.

manorom committed:
Add ompd_get_thread_handle for cuda.

* Code adapted from Marty Mcfadden's (<mcfadden8@llnl.gov>) code for the
  clang-ykt runtime.
* OMPD code in the nvptx device runtime library in libomptarget is
  currently included regardless of compile flags/defines

manorom committed:
[OMPD] Adds some support for omp states on cuda

* libompd can retrieve thread state information for a cuda thread
* ODB tries to summarize states of many cuda threads (this is still a
  bit ugly and doesnt repsect blocks)
* Currently openmp cuda threads can only be in an undefined state or
  parallel work state (parallel work is not consistent)

manorom committed:
[OMPD] Further align code to spec

* Moved some removed/renamed types out of ompd.h to ompd-private.h (we
  still need them internally i think)
* Removed dead code
* Some types have been renamed

Still need to rename all callbacks and check for completeness

manorom committed:
[OMPD] Aligned callback names with spec

manorom committed:
Fix formatting mistake

manorom committed:
[OMPD] Add parallel and reduction state to npvtx

manorom committed:
[OMPD] Add reduction state + save thread coords

manorom committed:
[OMPD] cleanup and + ompd parallel handle for cuda

manorom committed:
[OMPD] Remove unsupported api functions + fixes

manorom committed:
[OMPD] add support for some ICVs to libompd

manorom committed:
[OMPD] Add serial state in cuda device rtl

manorom committed:
[OMPD] Add some icvs + move values in ompd_types.h

manorom committed:
[OMPD] Align ompd.h formatting to spec

manorom committed:
[OMPD] changed some comments for current spec

manorom committed:
[OMPD] Some cleanup in gdb-wrapper

manorom committed:
[OMPD] fix omp version

manorom committed:
[OMPD] Add type compatibillity for new spec to odb

manorom committed:
[OMPD] Code clean-up in odb

manorom committed:
[OMPD] Fix bug introd. by type changes f. new spec

manorom committed:
[OMPD] Add "tasks" command to odb

manorom committed:
Merge remote-tracking branch 'origin/ompd' into ompd-devices

Conflicts:
	libompd/src/CMakeLists.txt
	libompd/src/omp-debug.cpp
	runtime/src/ompd-specific.cpp

manorom committed:
[OMPD] Add get_task_function and test code in odb

manorom committed:
[OMPD] Fix some formatting

manorom committed:
[OMPD] Fix formatting mistakes

manorom committed:
[OMPD] Fix linking for gdb-wrapper

manorom committed:
[OMPD] Save cuda kernel info in ompd handles

manorom committed:
Merge remote-tracking branch 'origin/master' into ompd-devices

manorom committed:
[OMPD] Set correct omp states for all modes (cuda)

manorom committed:
[OMPD] Fix thread handles in all exec modes (cuda)

manorom committed:
[OMPD] Add ompd break points for thread begin/end

manorom committed:
[OMPD] Add some support: parallel handles (cuda)

Adds support for ompd_get_current_parallel_handle and
ompd_get_enclosing_parallel_handle for cuda devices.

manorom committed:
[OMPD] remove usage of kernel id in odb

manorom committed:
[OMPD] Add ompd_get_thread_in_parallel for cuda

jprotze commented:
Some first comments: 
Please merge in the recent changes in the ompd branch to fix the merge conflict.
- the OMPD library should not use the global "new", but use the ompd callback for allocation (fixed in a recent commit on ompd) 
- Why is the if ompd_state deactivated in various functions? -> you need to set OMP_DEBUG=enabled (or OMP_OMPD=on, if it's not yet updated) and the variable should be set.
- Use your fix for the issue with the device_sizes.

manorom committed:
[OMPD] Move memory segment values to ompd_types.h

manorom commented:
Isn't the `new` operator overloaded to use the ompd allocation callback by deriving from `ompdAllocatable`?

manorom committed:
[OMPD] Re-enable checking for ompd state tracking

Also: remove get_*_data functions. Those are no longer part of the spec.

jprotze commented:
That was the idea when trying this. But we ran into issues with new just recently.

mcfadden8 commented on libompd/gdb-wrapper/ompd_typedefs.h.bak:
```diff
@@ -0,0 +1,521 @@
+/*
```
Was this file intentional?

manorom replied:
Nope. Removed with 300c218

manorom committed:
Remove accidentally committed .bak file

manorom committed:
Merge remote-tracking branch 'origin/ompd' into ompd-devices

Conflicts:
	libompd/src/TargetValue.cpp
	libompd/src/omp-debug.cpp

manorom committed:
Merge remote-tracking branch 'origin/ompd' into ompd-devices

Conflicts:
	runtime/src/ompd-specific.h

manorom committed:
[OMPD] Make odb initialization at first command

ODB now initializes the openmp process with ompd when the first ompd
commadn is issued, not when the debugger starts.
This is necessary because we now check for state tracking in the
process initialization function, and OpenMP runtime may not be present
when the debugger has just loaded the program.

manorom committed:
[OMPD] Add code to support some ICVs on cuda devs

manorom commented:
@jprotze resolved original conflict with e0538aa and c806b3e. Can someone take a look before we resolve the newest conflicts, or else we might and up with more merge commits than everything else?

manorom committed:
[OMPD] removed ompdAllocatable class

libompd does not use the new operator and instead uses the tool provides
malloc callback directly.

manorom committed:
[OMPD] Remove _ompd_device_handle_s struct

Its not part of the spec and not used by libompd.

manorom committed:
[OMPD] Work around for cuda-gdb name mangling bug

manorom committed:
[OMPD] Add support for cuda icvs to odb

manorom committed:
[OMPD] Fix enclosing_parallel + enable target ICVs

manorom committed:
[OMPD] Fix ompd_get_thread_in_parallel + add test

manorom committed:
[OMPD] Update ompd_types.h + remove kernelId field

manorom committed:
[OMPD] Fix ompd_types.h

manorom committed:
[OMPD] Add some support for tasks for cuda

manorom committed:
[OMPD] Fix parallel level for explicit tasks

manorom committed:
[OMPD] omp_device_t -> ompd_device_t

manorom committed:
[OMPD] Add examples to test ompd on cuda

manorom committed:
[OMPD] Add ompd breakpoints

manorom committed:
[OMPD] Fix task/parallel handle interaction

manorom committed:
[OMPD] Make OMPD configurable for deviceRTLs

manorom committed:
[OMPD] Clean up by removing unnecesarry header

manorom committed:
[OMPD} Remove deprecated ompd_process_handle_t

manorom committed:
[OMPD] Remove TODOs

manorom committed:
[OMPD] Add some kernel info + removed TODO notes

jprotze committed:
Add information on scheduling parent for the master task in a parallel region

Commit:
Initialize all address variables

jprotze committed:
Fix ompd_get_thread_id

Commit:
Initialize lwt field for ompd_get_task_in_parallel

jprotze committed:
Fix ompd_get_task_function for implicit tasks

Commit:
Use the right team to get the function pointer
